sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/EliminateJoinByCombineAggregate.scala

```diff
@@ -0,0 +1,196 @@
+        }
+      case (lp: Project, rp: Project) =>
+        val mergedInfo = mergePlan(lp.child, rp.child)
+        val mergedProjectList = ArrayBuffer[NamedExpression](lp.projectList: _*)
+
+        mergedInfo.map { case (newChild, outputMap, filters) =>
+          val allFilterReferences = filters.flatMap(_.references)
+          val newOutputMap = (rp.projectList ++ allFilterReferences).map { ne =>
+            val mapped = ne.transform {
+              case attr: Attribute =>
+                outputMap.getOrElse(attr.canonicalized, attr)
+            }.asInstanceOf[NamedExpression]
+
+            val withoutAlias = mapped match {
+              case Alias(child, _) => child
+              case e => e
+            }
+
+            val outputAttr = mergedProjectList.find {
+              case Alias(child, _) => child semanticEquals withoutAlias
+              case e => e semanticEquals withoutAlias
+            }.getOrElse {
+              mergedProjectList += mapped
+              mapped
+            }.toAttribute
+            ne.toAttribute.canonicalized -> outputAttr
+          }.toMap
+
+          (Project(mergedProjectList.toSeq, newChild), newOutputMap, filters)
+        }
+      // TODO support only one side contains Filter
+      case (lf: Filter, rf: Filter) =>
+        val mergedInfo = mergePlan(lf.child, rf.child)
+        mergedInfo.map { case (newChild, outputMap, _) =>
+          val rightCondition = rf.condition transform {
+            case attr: Attribute =>
+              outputMap.getOrElse(attr.canonicalized, attr)
+          }
+          val newCondition = Or(lf.condition, rightCondition)
+
+          (Filter(newCondition, newChild), outputMap, Seq(lf.condition, rightCondition))
+        }
+      case (ll: LeafNode, rl: LeafNode) =>
+        if (ll.canonicalized == rl.canonicalized) {
+          val outputMap = rl.output.zip(ll.output).map { case (ra, la) =>
+            ra.canonicalized -> la
+          }.toMap
+
+          Some((ll, outputMap, Seq.empty))
+        } else {
+          None
+        }
+      case (ls: SerializeFromObject, rs: SerializeFromObject) =>
+        if (ls.canonicalized == rs.canonicalized) {
+          val outputMap = rs.output.zip(ls.output).map { case (ra, la) =>
+            ra.canonicalized -> la
+          }.toMap
+
+          Some((ls, outputMap, Seq.empty))
+        } else {
+          None
+        }
+      case _ => None
+    }
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!conf.eliminateJoinByCombineAggregateEnabled) return plan
+
+    plan.transformDownWithPruning(_.containsAnyPattern(JOIN, AGGREGATE), ruleId) {
+      case j @ Join(_, _, joinType, None, _) if isSupportedJoinType(joinType) =>
+        val aggregates = ArrayBuffer.empty[Aggregate]
+        if (collectAggregate(j, aggregates)) {
+          var finalAggregate: Option[LogicalPlan] = None
+          for ((aggregate, i) <- aggregates.tail.zipWithIndex
```

Comment:
Seems like you are trying to merge aggregates in sequential order only. What if the first can't be merged with the rest, but the rest can be merged together?

Reply:
This rule can merge any continuous part of to be merged aggregates in sequential order.
If the first can't be merged with the rest, the rest can still be merged together.

Reply:
Got it now thanks! Maybe we could try merging an aggregate with all the previous ones that couldn't be merged together.

Reply:
But I'm not sure how can you replace the `Join` to the last merged aggregate only?